staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go

```diff
@@ -284,10 +287,43 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
 	return &authenticator.Response{
 		User: &user.DefaultInfo{
 			Name:   chain[0].Subject.CommonName,
+			UID:    parseUIDFromCert(chain[0]),
 			Groups: chain[0].Subject.Organization,
 			Extra: map[string][]string{
 				user.CredentialIDKey: {id},
 			},
 		},
 	}, true, nil
 })
+
+// Using CNCF PEN for Kubernetes:
+// https://www.iana.org/assignments/enterprise-numbers.txt
+//
+// 57683
+//
+//	Cloud Native Computing Foundation
+//	  Davanum Srinivas
+//	    sc1&kubernetes.io
+var uidOID = asn1.ObjectIdentifier([]int{1, 3, 6, 1, 4, 1, 57683, 2})
+
+func parseUIDFromCert(cert *x509.Certificate) string {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.AllowParsingUserUIDFromCertAuth) {
+		return ""
+	}
+
+	uid := ""
+	found := false
+	for _, name := range cert.Subject.Names {
+		if !name.Type.Equal(uidOID) {
+			continue
+		}
+		if found {
+			// Require that our OID is only used once.
+			// If there are multiple, we leave UID unset.
+			return ""
+		}
+		found = true
+		uid, _ = name.Value.(string)
```

Comment:
Updated this to return an error if string parsing isn't successful 👍 

I was assuming that strings were always parsed as such, but I'll need to verify. I'll dig into the asn1 parsing, and whether we can surface a non-string in a test 🤔

Reply:
> is there a way to exercise the non-string case in a test?

Yes, this is possible, but I'm not sure if it's worth the effort. Here's a commit that tests the behavior: https://github.com/kubernetes/kubernetes/pull/127897/commits/ef3775610ff70aa23b056d1d03e1f2da43c805c4

but I rolled it back because I don't think it's the proper way to "edit" a cert. Instead, we'd need to generate a new cert after setting the binary field.

> I'm fuzzy on asn1 parsing, do strings always show up as strings or can they be []byte depending on the encoder?

Great question. Yes, I believe it's possible parse a cert with a []byte as a field. I'd love to prove this in a test, now that we've added the error handling. I'm also willing to do it as a followup, if we want to merge this PR sooner.

Reply:
After more investigation, I don't think it's possible to generate a cert with a non-string value.

Some details are in [my latest comment](https://github.com/kubernetes/kubernetes/pull/127897#issuecomment-2467209437)

Sorry for the back-and-forth!